packages/jsii-pacmak/lib/targets/golang/types/interface.ts

```diff
@@ -1,55 +1,102 @@
+  public emit(code: CodeMaker) {
+    const name = this.property.name;
+    const type =
+      this?.references?.scopedName(this.parent.parent) ??
+      this.property.toString();
+
+    code.line(`Get${name}() ${type}`);
+    if (!this.property.protected) {
+      code.line(`Set${name}()`);
+    }
   }
 }
 
-export class Interface extends GoType {
-  public constructor(public type: InterfaceType) {
-    super(type);
+export class InterfaceMethod implements InterfaceField {
+  public readonly name: string;
+  public readonly references?: GoTypeRef;
+  public constructor(public parent: Interface, public readonly method: Method) {
+    this.name = this.method.name;
+
+    if (this.method.returns.type) {
+      this.references = new GoTypeRef(
+        parent.parent.root,
+        this.method.returns.type,
+      );
+    }
   }
 
-  public emit(code: CodeMaker): void {
-    code.openBlock(`type ${this.localName} interface`);
-    Object.values(this.type.getMethods()).forEach((method) =>
-      this.emitInterfaceMethod(code, method),
+  public emit(code: CodeMaker) {
+    const returns = this.method.returns.type.void
+      ? ''
+      : this?.references?.scopedName(this.parent.parent) ??
+        this.method.returns.toString();
+    code.line(`${code.toPascalCase(this.name)}()${returns}`);
+  }
+}
+
+export class Interface extends GoType implements GoEmitter {
+  public readonly properties: InterfaceProperty[];
+  public readonly methods: InterfaceMethod[];
+
+  public constructor(parent: Module, public type: InterfaceType) {
+    super(parent, type);
+
+    this.properties = Object.values(this.type.getProperties()).map(
+      (prop) => new InterfaceProperty(this, prop),
     );
 
-    Object.values(this.type.getProperties()).forEach((property) =>
-      this.emitInterfaceProperty(code, property),
+    this.methods = Object.values(this.type.getMethods()).map(
+      (method) => new InterfaceMethod(this, method),
     );
-    code.closeBlock();
-    code.line();
   }
 
-  private emitInterfaceProperty(code: CodeMaker, property: Property) {
-    const propName = code.toPascalCase(property.name);
-    const type = new TypeMapper(property.type).mapType();
-
-    code.line(`Get${propName}() ${type}`);
-    // if (!property.protected) {
-    //   code.line(`set${name}()`);
-    // }
-  }
+  public emit(code: CodeMaker) {
+    code.openBlock(`type ${this.name} interface`);
 
-  private emitInterfaceMethod(code: CodeMaker, method: Method) {
-    const returns = method.returns.type.void
-      ? ''
-      : ` ${new TypeMapper(method.returns.type).mapType()}`;
+    this.properties.forEach((property) => property.emit(code));
+    this.methods.forEach((method) => method.emit(code));
 
-    const methodName = code.toPascalCase(method.name);
+    code.closeBlock();
+    code.line();
+  }
 
-    code.line(`${methodName}()${returns}`);
+  public get dependencies(): Module[] {
+    return [
+      ...this.properties.reduce((accum: Module[], property) => {
```

Comment:
Personally, I find this functional style pretty hard to read -- is this idiomatic elsewhere in this repo? @RomainMuller